libraries/app.rb

```diff
@@ -0,0 +1,150 @@
+        end
+
+        def install_app
+          # If we force the install remove target folder if it exists.
+          directory install_dir do
+            recursive true
+            action :delete
+          end if new_resource.force && exist?
+
+          if new_resource.url.end_with?('dmg')
+            install_dmg
+          else
+            raise 'Unsupported package provided Xcode installer must be provided as a DMG'
+          end
+
+          post_install
+          accept_eula
+          delete_installer_file if exist?
+        end
+
+        def install_dmg
+          dmg_package new_resource.app do
+            source new_resource.url
+            checksum new_resource.checksum
+            dmg_name ::File.basename(new_resource.url, '.dmg')
+            owner 'root'
+            type 'app'
+            action :install
+          end
+        end
+
+        # The `post_install` function is only called if the function 'install_app' is called.
+        def post_install
+          ruby_block 'Xcode post install' do
+            block do
+              if new_resource.multi_install || !new_resource.install_root.eql?('/Applications')
+                ::File.rename('/Applications/Xcode.app', install_dir)
+              end
+              # Install any additional packages hiding in the Xcode installation path
+              xcode_packages =
+                ::Dir.entries("#{install_dir}/Contents/Resources/Packages/").select { |f| !::File.directory? f }
+              xcode_packages.each do |pkg|
+                execute "Installing Xcode package [#{pkg}]" do
+                  command "sudo installer -pkg #{install_dir}/Contents/Resources/Packages/#{pkg} -target /"
+                end
+              end unless xcode_packages.nil?
+              execute 'xcode-select' do
+                command "xcode-select -s #{install_dir}/Contents/Developer"
+              end
+            end
+            only_if { ::Dir.exist?('/Applications/Xcode.app') }
+          end
+        end
+
+        # The `accept_eula` function is only called if the function 'install_app' is called.
+        def accept_eula
+          # Accept the Xcode license, this creates the /Library/Preferences/com.apple.dt.Xcode.plist file
+          execute 'Accept xcode license' do
+            command "#{install_dir}/Contents/Developer/usr/bin/xcodebuild -license accept"
+            action :run
+            only_if { ::File.exist?("#{install_dir}/Contents/Developer/usr/bin/xcodebuild") }
+            # TODO: Do a proper version check such that the oldest version is always accepted.
+            # only_if { newer?("#{install_dir}/Contents/Developer/usr/bin/xcodebuild -version") }
+          end
+        end
+
+        def install_dir
+          ::File.join(new_resource.install_root,
+                      "Xcode#{new_resource.multi_install ? "_#{new_resource.version}" : ''}.app")
+        end
+
+        def exist?
+          new_resource.force ? false : Dir.exist?(install_dir)
```

Comment:
When force is true, this returns false, which causes a problem for action `uninstall` because it says it does not exist, so doesn't do uninstall.